File: antlropt/org/perses/antlr/ast/RuleNameRegistry.java
/*
 * Copyright (C) 2018-2020 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.antlr.ast;

import com.google.common.base.MoreObjects;
import com.google.common.base.Preconditions;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.perses.antlr.RuleType;

import java.util.Optional;
import java.util.WeakHashMap;

public class RuleNameRegistry {

  private final WeakHashMap<String, RuleNameHandle> ruleNameMap = new WeakHashMap<>();
  private int idGenerator;

  public RuleNameHandle getOrCreate(String ruleName) {
    RuleNameHandle value = ruleNameMap.get(ruleName);
    if (value == null) {
      value = createOrThrow(ruleName);
    }
    Preconditions.checkState(ruleName.equals(value.ruleName));
    Preconditions.checkState(ruleNameMap.containsKey(ruleName));
    Preconditions.checkState(ruleNameMap.containsValue(value));
    return value;
  }

  public RuleNameHandle createOrThrow(String ruleName) {
    return createOrThrow(ruleName, /*origin=*/ null);
  }

  private RuleNameHandle createOrThrow(String ruleName, @Nullable RuleNameHandle origin) {
    Preconditions.checkArgument(!ruleNameMap.containsKey(ruleName));
    RuleNameHandle value = new RuleNameHandle(ruleName, idGenerator++, origin);
    ruleNameMap.put(ruleName, value);
    return value;
  }

  public Optional<RuleNameHandle> get(String ruleName) {
    return Optional.ofNullable(ruleNameMap.get(ruleName));
  }

  public RuleNameHandle getOrThrow(String ruleName) {
    RuleNameHandle value = ruleNameMap.get(ruleName);
    Preconditions.checkNotNull(value, "no rule '%s' in registry", ruleName);
    return value;
  }

  /** This is just a handle to the rule name. The rule name can be changed. */
  public final class RuleNameHandle implements Comparable<RuleNameHandle> {
    private final String ruleName;
    private final int id;
    private final RuleType type;
    @Nullable private final RuleNameHandle origin;
    private int auxiliaryIdGenerator = 1;

    private RuleNameHandle(String ruleName, int id, @Nullable RuleNameHandle origin) {
      this.ruleName = ruleName;
      this.id = id;
      type = RuleType.determineType(ruleName);
      this.origin = origin;
    }

    public String get() {
      return ruleName;
    }

    public int getId() {
      return id;
    }

    public RuleType getType() {
      return type;
    }

    public RuleNameHandle createAuxiliaryRuleName(RuleType ruleType) {
      if (origin != null) {
        return origin.createAuxiliaryRuleName(ruleType);
      }
      String prefix = ruleType == type ? "" : getDefaultAuxiliaryPrefix(ruleType);
      return createOrThrow(prefix + ruleName + '_' + auxiliaryIdGenerator++, this);
    }

    private String getDefaultAuxiliaryPrefix(RuleType type) {
      switch (type) {
        case KLEENE_PLUS:
        case KLEENE_STAR:
        case OPTIONAL:
        case ALT_BLOCKS:
          return type.getSignaturePrefix();
        case TOKEN:
          return "AUXILIARY__";
        case OTHER_RULE:
          return "auxiliary__";
        default:
          throw new RuntimeException("Cannot reach here");
      }
    }

    @Override
    public int compareTo(RuleNameHandle other) {
      final RuleNameHandle origin = MoreObjects.firstNonNull(this.origin, this);
      final RuleNameHandle otherOrigin = MoreObjects.firstNonNull(other.origin, other);
      final int originCmp = Integer.compare(origin.id, otherOrigin.id);
      if (originCmp != 0) {
        return originCmp;
      }
      return Integer.compare(id, other.id);
    }

    @Override
    public String toString() {
      return MoreObjects.toStringHelper(this).add("id", id).add("name", ruleName).toString();
    }
  }
}

File: antlropt/src/org/perses/antlr/ast/RuleNameRegistry.kt
/*
 * Copyright (C) 2018-2024 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.antlr.ast

import com.google.common.base.MoreObjects
import org.perses.antlr.RuleType
import java.util.WeakHashMap

class RuleNameRegistry {

  private val ruleNameMap = WeakHashMap<String, RuleNameHandle>()

  private var idGenerator = 0

  fun getOrCreate(ruleName: String): RuleNameHandle {
    var value = ruleNameMap[ruleName]
    if (value == null) {
      value = createOrThrow(ruleName)
    }
    check(ruleName == value.ruleName)
    check(ruleNameMap.containsKey(ruleName))
    check(ruleNameMap.containsValue(value))
    return value
  }

  fun createOrThrow(ruleName: String): RuleNameHandle {
    return createOrThrow(ruleName, origin = null)
  }

  private fun createOrThrow(ruleName: String, origin: RuleNameHandle?): RuleNameHandle {
    require(!ruleNameMap.containsKey(ruleName))
    val value = RuleNameHandle(ruleName, idGenerator++, origin)
    ruleNameMap[ruleName] = value
    return value
  }

  operator fun get(ruleName: String): RuleNameHandle? {
    return ruleNameMap[ruleName]
  }

  fun getOrThrow(ruleName: String): RuleNameHandle {
    val value = ruleNameMap[ruleName]
    check(value != null) { "no rule $ruleName in registry" }
    return value
  }

  /** This is just a handle to the rule name. The rule name can be changed.  */
  inner class RuleNameHandle internal constructor(
    val ruleName: String,
    val id: Int,
    val origin: RuleNameHandle?,
  ) : Comparable<RuleNameHandle> {

    private var auxiliaryIdGenerator = 1

    fun createAuxiliaryRuleName(ruleType: RuleType): RuleNameHandle {
      if (origin != null) {
        return origin.createAuxiliaryRuleName(ruleType)
      }
      val suffix = "${ruleName}_${auxiliaryIdGenerator++}"
      return createOrThrow(ruleType.addPrefixIfMissing(suffix), this)
    }

    override fun compareTo(other: RuleNameHandle): Int {
      val origin = MoreObjects.firstNonNull(origin, this)
      val otherOrigin = MoreObjects.firstNonNull(other.origin, other)
      val originCmp = Integer.compare(origin.id, otherOrigin.id)
      return if (originCmp != 0) {
        originCmp
      } else {
        Integer.compare(id, other.id)
      }
    }

    override fun toString(): String {
      return MoreObjects.toStringHelper(this).add("id", id).add("name", ruleName).toString()
    }
  }
}
